### PR TITLE
Add ability for logs to be downloaded. Adds #1968

### PR DESCRIPTION
**Pull request by lemonsked**
This adds the ability for logs to be downloaded via the HTTPD.

### CODE REVIEW

File: src/main/java/me/totalfreedom/totalfreedommod/httpd/module/Module_logfile.java [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    public String title()
+    {
+        return "TotalFreedomMod :: Logfiles";
+    }
+
+    public String body() throws ResponseOverrideException
+    {
+        if (!LOG_FOLDER.exists())
+        {
+            return HTMLGenerationTools.paragraph("Can't find the logs folder.");
```
Comment (Wild1145) [inline src/main/java/me/totalfreedom/totalfreedommod/httpd/module/Module_logfile.java:59]: Should this not also log out as an error to the server?

File: src/main/java/me/totalfreedom/totalfreedommod/httpd/module/Module_logfile.java [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+        final String[] args = StringUtils.split(uri, "/");
+        final ModuleMode mode = ModuleMode.getMode(getArg(args, 1));
+
+        switch (mode)
+        {
+            case LIST:
+            {
+                if (!isAuthorized(remoteAddress))
+                {
+                    out.append(HTMLGenerationTools.paragraph("Log files access denied: Your IP, " + remoteAddress + ", is not registered to a superadmin on this server."));
```
Comment (Wild1145) [inline src/main/java/me/totalfreedom/totalfreedommod/httpd/module/Module_logfile.java:73]: This should also be logged server side as well to keep a log of who is attempting to access the logs.

File: src/main/java/me/totalfreedom/totalfreedommod/httpd/module/Module_logfile.java [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+                            .append(StringUtils.join(LogFilesFormatted, "\r\n"))
+                            .append("</ul>");
+                }
+                break;
+            }
+            case DOWNLOAD:
+            {
+                if (!isAuthorized(remoteAddress))
+                {
+                    out.append(HTMLGenerationTools.paragraph("Log files access denied: Your IP, " + remoteAddress + ", is not registered to a superadmin on this server."));
```
Comment (Wild1145) [inline src/main/java/me/totalfreedom/totalfreedommod/httpd/module/Module_logfile.java:110]: Once again this should be logged to the server console as well.

**Review by Wild1145 — CHANGES_REQUESTED** [review]

**Comment by lemonsked** [comment]
It now logs it to the server console too.
I didn't first make it log to the server console since if you spam refresh the page, it'll spam the logs.

**Review by Wild1145 — APPROVED** [review]

**Comment by ghost** [comment]
You can already download logs via HTTPD @OxLemonxO ....

**Comment by lemonsked** [comment]
@Commodore64x No, can you download it like this?
http://i.imgur.com/nZTy5t9.png

**Comment by ghost** [comment]
but it will need to be wiped every week to free space for server?

**Comment by lemonsked** [comment]
....... @marcocorriero, all the files that are stored in /logs are simply just listed and download links are shown

File: src/main/java/me/totalfreedom/totalfreedommod/httpd/module/Module_logfile.java [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+    {
+        return "TotalFreedomMod :: Logfiles";
+    }
+
+    public String body() throws ResponseOverrideException
+    {
+        if (!LOG_FOLDER.exists())
+        {
+            return HTMLGenerationTools.paragraph("Can't find the logs folder.");
+            FLog.warning("The logfile module failed to find the logs folder.");
```
Comment (JeromSar) [inline src/main/java/me/totalfreedom/totalfreedommod/httpd/module/Module_logfile.java:58]: Unreachable code.

File: src/main/java/me/totalfreedom/totalfreedommod/httpd/module/Module_logfile.java [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+        final ModuleMode mode = ModuleMode.getMode(getArg(args, 1));
+
+        switch (mode)
+        {
+            case LIST:
+            {
+                if (!isAuthorized(remoteAddress))
+                {
+                    out.append(HTMLGenerationTools.paragraph("Log files access denied: Your IP, " + remoteAddress + ", is not registered to a superadmin on this server."));
+                    FLog.info("An unregistered IP (" + remoteAddress ") has tried to access the log files");
```
Comment (JeromSar) [inline src/main/java/me/totalfreedom/totalfreedommod/httpd/module/Module_logfile.java]: This is not valid Java, I believe.

File: src/main/java/me/totalfreedom/totalfreedommod/httpd/module/Module_logfile.java [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+                            .append("</ul>");
+                }
+                break;
+            }
+            case DOWNLOAD:
+            {
+                if (!isAuthorized(remoteAddress))
+                {
+                    out.append(HTMLGenerationTools.paragraph("Log files access denied: Your IP, " + remoteAddress + ", is not registered to a superadmin on this server."));
+                    FLog.info("An unregistered IP (" + remoteAddress ") has tried to download a log file");
```
Comment (JeromSar) [inline src/main/java/me/totalfreedom/totalfreedommod/httpd/module/Module_logfile.java]: This is not valid Java, I believe.

**Review by JeromSar — CHANGES_REQUESTED** [review]

**Comment by lemonsked** [comment]
Fixed @JeromSar

**Review by JeromSar — APPROVED** [review]

**Comment by JeromSar** [comment]
Just needs testing now. :)

**Comment by Wild1145** [comment]
Tested and happy. One minor typo which I've pushed a fix for. Merging as resolved. Thanks for the contribution @OxLemonxO and apologies it's taken so long for me to get around to merging this.

**Pull request merged**